npeps/npep-182-cidr-object-peer.md

```diff
@@ -0,0 +1,195 @@
+# NPEP-182: Add new CIDR object peer for northbound traffic
+
+* Issue: [#182](https://github.com/kubernetes-sigs/network-policy-api/issues/182)
+* Status: Provisional
+
+## Co-Authors
+@joestringer and @networkop for raising relevant user stories
+
+## TLDR
+
+This NPEP proposes adding support for a new CIDRGroup object peer type for
+cluster egress (northbound) traffic control that can be referred in the
+`AdminNetworkPolicy` and `BaselineAdminNetworkPolicy` API objects using selectors.
+[NPEP-126](https://network-policy-api.sigs.k8s.io/npeps/npep-126-egress-traffic-control/#implementing-egress-traffic-control-towards-cidrs)
+already adds support for inline CIDR peer type directly on the
+`AdminNetworkPolicy` and `BaselineAdminNetworkPolicy` API objects. This NPEP proposes
+adding more extensibility by introducing a new CIDRGroup object in addition to the
+inline CIDR peers so that users can choose either of these methods based on their needs.
+
+## Goals
+
+* Provide users with a way to group their CIDRs in a meaningful
+manner which can then be referred to from ANP and BANP objects.
+
+## Non-Goals
+
+## Introduction
+
+The current approach of defining inline CIDR peers works well
+if the number of CIDR blocks involved in defining policies are
+less in number and mostly static in nature. However in environments
+where we could have a more dynamic setup, the management of inline CIDR
+peers gets more tricker an cumbersome. In such cases having a way to
+group CIDR blocks together to represent an entity or a group of
+entities which the policy can refer to as a network peer can be useful.
+This also ensures reference of same CIDR group peer from ANP and BANP
+stays consistent and any changes to the list of CIDR blocks only involves
+editing the object itself and not the rules in the policy.
+
+### User Stories for CIDRGrouping
+
+* As a cluster admin I want to be able to create admin network policies that
+match a dynamic set of external IPs (e.g. set of VMs or set of directly reachable
+Pods in another cluster). I may not be able to use FQDN rules for that due to
+TTL being too long or simply lack of DNS service discovery in an external system.
+As a cluster admin, I would create CIDR group resource and a BGP controller that
+would manage it. The mapping between BGP communities and CIDR group resource names
+is a BGP controller configuration (e.g. annotation on the CIDR group resource).
+The speed of IP churn is bounded by the BGP advertisement interval and can be
+further reduced by the BGP controller implementation.
+
+* As a cluster administrator I want to to ensure that pods can reach
```

Comment:
Just playing devil's advocate with maybe a really bad idea:

What if we try to use the EndpointSlice API to solve this set of problems? 

Pros:
- Fewer network policy objects
- Make headway into defining the Service selector API

Cons:
- Abusing(?) the EndpointSlice idiom

Reply:
@rahulkjoshi can you specify IP prefixes or only IP addresses in EndpointSlice?

Reply:
😦 yeah you're right only individual addresses in EndpointSlice

My underlying thought with this situation was whether there's overlap with a more general service discovery problem. My thinking was that whatever application the rule protects also needs to discover the IPs it's talking to.

I was hoping that plugging into that system might be more ergonomic? But it's more likely to just be incredibly complicated and not easily generalizable.